module/os/freebsd/zfs/vdev_file.c

```diff
@@ -260,16 +260,9 @@ vdev_file_io_start(zio_t *zio)
 		zio_execute(zio);
 		return;
 	} else if (zio->io_type == ZIO_TYPE_TRIM) {
-#ifdef notyet
-		int mode = 0;
-
 		ASSERT3U(zio->io_size, !=, 0);
-
-		/* XXX FreeBSD has no fallocate routine in file ops */
-		zio->io_error = zfs_file_fallocate(vf->vf_file,
-		    mode, zio->io_offset, zio->io_size);
-#endif
-		zio->io_error = SET_ERROR(ENOTSUP);
+		zio->io_error = zfs_file_deallocate(vf->vf_file,
+		    zio->io_offset, zio->io_size);
 		zio_execute(zio);
```

Comment:
I wonder why the vdev_file code on both FreeBSD and Linux runs TRIM directly, not in taskq context same as read/write.  We've recently hit and fixed the issue when on Linux sync TRIM caused performance problems by blocking ZIO pipeline for no good reason.

Reply:
Mm, also flush, except in that weird PF_FSTRANS case. Is there any reason not to use `vdev_file_taskq` for everything? Seems like it would be fine.

Reply:
I can't think of anything except it was an attempt to save on context switches, which may be counterproductive if multiple TRIMs or flushes are to be executed same time.

Reply:
Mm. Sounds like worth trying then, but I'll come back to it in another PR because it needs a bt more thought and testing.